apps/stack/src/config/create-event.ts

```diff
@@ -56,12 +55,58 @@ export function createEvent<
   return {
     publish,
     source,
-    type,
+    detailType,
     shape: {
-      properties: {} as Properties<Shape>,
-      metadata: {} as Parameters<Publish<Shape, MetadataShape>>[1],
-      metadataFn: undefined,
+      metadata,
+      properties
     },
   };
 }
 
+export type EventDefinition<
+  Source extends string,
+  DetailType extends string,
+  PropertiesShape extends z.ZodRawShape,
+  MetadataShape extends z.ZodRawShape,
+> = {
+  source: Source,
+  detailType: DetailType
+  shape: {
+    properties: PropertiesShape,
+    metadata: MetadataShape
+  }
+}
+
+type EventPayload<
+  PropertiesShape extends z.ZodRawShape,
+  MetadataShape extends z.ZodRawShape,
+> = {
+  properties: InferShapeOutput<PropertiesShape>;
+  metadata: InferShapeOutput<MetadataShape>;
+}
+
+export const EventHandler = <
+  Source extends string,
+  DetailType extends string,
+  PropertiesShape extends z.ZodRawShape,
+  MetadataShape extends z.ZodRawShape>(
+    event: EventDefinition<Source, DetailType, PropertiesShape, MetadataShape>,
+    cb: (ev: EventPayload<PropertiesShape,MetadataShape>) => Promise<void>) => {
+  const { source: targetSource, detailType: targetDetailType } = event;
+  const eventSchema = z.object({
+    properties: z.object(event.shape.properties),
+    metadata: z.object(event.shape.metadata),
+  });
+
+  return async (
+    event: EventBridgeEvent<string, unknown>
+  ) => {
+    if (event["detail-type"] !== targetDetailType || event.source !== targetSource) {
+      console.warn(`Warning: Invalid event handler configuration, expected event ${targetSource}.${targetDetailType} but got ${event.source}.${event["detail-type"]}`);
+    }
+    const parseResult = eventSchema.safeParse(event.detail);
+    if (!parseResult.success) return console.error(`ERROR: Failed to parse event detail '${targetSource}.${targetDetailType}'. Reason: ${parseResult.error}`);
```

Comment:
Return here because is not a retryable error. Not sure if we can fail the lambda as a non-retryable error